### PR TITLE
Add Details about API Guarantees Surrounding grpc_arg_pointer_vtable

### PR DESCRIPTION
**Pull request by ncteisen**

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: include/grpc/grpc.h [hunk 1]
```diff
@@ -269,9 +269,11 @@ GRPCAPI void grpc_channel_get_info(grpc_channel* channel,
 
 /** Create a client channel to 'target'. Additional channel level configuration
     MAY be provided by grpc_channel_args, though the expectation is that most
-    clients will want to simply pass NULL. See grpc_channel_args definition for
-    more on this. The data in 'args' need only live through the invocation of
-    this function. */
+    clients will want to simply pass NULL. The user data in 'args' need only
+    live through the invocation of this function. However, if any args of the
+    'ponter' type are passed, then the referenced vtable must be maintained
```
Comment (nathanielmanistaatgoogle) [inline include/grpc/grpc.h]: Typo "ponter".

File: include/grpc/grpc.h [hunk 2]
```diff
@@ -364,8 +366,11 @@ GRPCAPI grpc_call_error grpc_server_request_registered_call(
 
 /** Create a server. Additional configuration for each incoming channel can
     be specified with args. If no additional configuration is needed, args can
-    be NULL. See grpc_channel_args for more. The data in 'args' need only live
-    through the invocation of this function. */
+    be NULL. The user data in 'args' need only live through the invocation of
+    this function. However, if any args of the 'ponter' type are passed, then
```
Comment (nathanielmanistaatgoogle) [inline include/grpc/grpc.h]: "ponter" here too.

File: include/grpc/grpc_security.h [hunk 3]
```diff
@@ -300,7 +300,13 @@ GRPCAPI grpc_call_credentials* grpc_metadata_credentials_create_from_plugin(
 
 /** --- Secure channel creation. --- */
 
-/** Creates a secure channel using the passed-in credentials. */
+/** Creates a secure channel using the passed-in credentials. Additional
+    channel level configuration MAY be provided by grpc_channel_args, though
+    the expectation is that most clients will want to simply pass NULL. The
+    user data in 'args' need only live through the invocation of this function.
+    However, if any args of the 'ponter' type are passed, then the referenced
```
Comment (nathanielmanistaatgoogle) [inline include/grpc/grpc_security.h]: ... and "ponter" here too.

File: include/grpc/impl/codegen/grpc_types.h [hunk 4]
```diff
@@ -119,9 +119,13 @@ typedef struct {
     These configuration options are modelled as key-value pairs as defined
     by grpc_arg; keys are strings to allow easy backwards-compatible extension
     by arbitrary parties. All evaluation is performed at channel creation
-    time (i.e. the values in this structure need only live through the
+    time (i.e. the keys and values in this structure need only live through the
     creation invocation).
 
+    However, if one of the args has grpc_arg_type==GRPC_ARG_POINTER, then the
+    grpc_arg_pointer_vtable must live until the channel args are done being
+    used by core (i.e. when the object they were passed to is destroyed).
```
Comment (nathanielmanistaatgoogle) [inline include/grpc/impl/codegen/grpc_types.h]: "the object for use with which they were passed"? One doesn't pass channel args to a channel; one passes channel args to the channel creation function.

File: include/grpc/grpc.h (hunk 1 above)
Comment (ncteisen) [reply include/grpc/grpc.h]: Oops done

File: include/grpc/grpc.h (hunk 2 above)
Comment (ncteisen) [reply include/grpc/grpc.h]: Done

File: include/grpc/grpc_security.h (hunk 3 above)
Comment (ncteisen) [reply include/grpc/grpc_security.h]: Done

File: include/grpc/impl/codegen/grpc_types.h (hunk 4 above)
Comment (ncteisen) [reply include/grpc/impl/codegen/grpc_types.h]: Done

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by ncteisen** [comment]
Squashed

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**